### PR TITLE
Simplify rules_nixpkgs_core module extensions

### PR DESCRIPTION
**Pull request by aherrmann**
Part of #183.

The previous implementation of the rules_nixpkgs_core module extensions (https://github.com/tweag/rules_nixpkgs/pull/344, https://github.com/tweag/rules_nixpkgs/pull/363, https://github.com/tweag/rules_nixpkgs/pull/369, https://github.com/tweag/rules_nixpkgs/pull/364) used a [hub-repository](https://github.com/tweag/rules_nixpkgs/pull/364) to manage Nix repository and package import scope.

Since then, Bazel (>= 6.3) has introduced [isolated module extension instances](https://github.com/bazelbuild/bazel/pull/18529) and [automatic generation of `use_repo` stanzas](https://docs.google.com/document/d/1dj8SN5L6nwhNOufNqjBhYkk5f-BJI_FPYWKxlB3GAmA/edit?disco=AAAArdGBwhc). Both these features solve the same problems that the hub-repository implementation addressed, see below.

This PR replaces the previous hub-repository based implementation by a simpler implementation based on these new Bazel features. A clear advantage is that this means that users can use plain Bazel labels to refer to imported repositories and packages instead of having to access them through hub-repository accessors. This should also be an advantage in terms of WORKSPACE to bzlmod migration effort.

How isolated extensions and automatically generated `use_repo` address rules_nixpkgs's requirements:
- Separating globally unified and locally specialized packages. New: Globally unified behavior is the default, and locally specialized behavior is achieved using an isoalted module extension instance.
- Exposing imports under human friendly names. New: Because isolated module extension instances have their own namespace, we no longer need to worry about name mangling ourselves. Instead, we can generate external repositories with human friendly names directly, either the canonical attribute name in case of globally unified packages, or a human defined name in case of isolated instances or overrides in root or rules_nixpkgs_core. Thanks to the automatically generated `use_repo` stanzas, it is very easy for a user to manage the imports.

### CODE REVIEW

**Comment by aherrmann** [comment]
The [second CI failure](https://github.com/tweag/rules_nixpkgs/actions/runs/6390957120/job/17345262106?pr=423#step:6:846) seems to be caused by an incompatibility with more recent nixpkgs revisions and MacOS 11 on x86_64. I have not been able to track down the exact reason, but worked around the issue by falling back to `pkgs.coreutils`. I've opened https://github.com/tweag/rules_nixpkgs/issues/424 to track this.

File: flake.nix [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   inputs = {
-    nixpkgs.url = "github:NixOS/nixpkgs/nixos-22.11";
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
```
Comment (layus) [inline flake.nix:3]: Any specific reason for this change ?

File: testing/core/tests/nixpkgs_repositories.bzl [hunk 2]
```diff
@@ -10,29 +8,25 @@ load(
...
-        file_nixpkgs = nix_repo("rules_nixpkgs_core_testing", "file_nixpkgs")
-        nixpkgs_content = nix_repo("rules_nixpkgs_core_testing", "nixpkgs_content")
-    else:
-        nixpkgs = "@nixpkgs"
+    nixpkgs = "@nixpkgs"
+    remote_nixpkgs = "@remote_nixpkgs"
+    http_nixpkgs = "@http_nixpkgs"
+    file_nixpkgs = "@file_nixpkgs"
+    nixpkgs_content = "@nixpkgs_content"
+    if not bzlmod:
```
Comment (layus) [inline testing/core/tests/nixpkgs_repositories.bzl:16]: XXX: Are all these "global" names defined in the documentation ?

File: tests/BUILD.bazel [hunk 3]
```diff
@@ -8,7 +8,7 @@ sh_test(
         "//nixpkgs:srcs",
         "//tests/invalid_nixpkgs_package:srcs",
         "@coreutils_static//:bin",
-        "@nix_2_7//:bin",
+        "@nix_2_10//:bin",
```
Comment (layus) [inline tests/BUILD.bazel:11]: Any reason for thins change, except general maintenance ?

**Review by layus — APPROVED** [review]
Interesting changeset. Overall not much changed in the interface, except for some simplifications, so this is a much welcome change.

I would love to see the implications for more complex use cases of rules_nixpkgs, where @nixpkgs is not a vanilla upstream nixpkgs, but I think there is kind of a global assumption that @nixpkgs is some version of upstream nixpkgs.

I am still annoyed that bzlmod (at least for rules_nixpkgs) relies on so many conventions and unchecked assumptions, but there is not much we can do about that.

Some comments I left are merely notes for me to check after the review, these are marked with XXX.

File: flake.nix (hunk 1 above)
Comment (aherrmann) [reply flake.nix:3]: Yes, to get access to Bazel version 6.3.2, which is not yet available in NixOS-22.11.

File: tests/BUILD.bazel (hunk 3 above)
Comment (aherrmann) [reply tests/BUILD.bazel:11]: Yes, version 2.7 (and 2.8 and 2.9) are no longer available in the more recent nixpkgs revision (unstable). So we need to bump the version.

**Pull request merged**